pkg/detectors/snowflake/snowflake.go

```diff
@@ -0,0 +1,158 @@
+			hasNumber = true
+		}
+
+		if hasLower && hasUpper && hasNumber && minLen {
+			return password, true
+		}
+	}
+
+	return "", false
+}
+
+// FromData will find and optionally verify Snowflake secrets in a given set of bytes.
+func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	dataStr := string(data)
+
+	accountMatches := accountIdentifierPat.FindAllStringSubmatch(dataStr, -1)
+
+	usernameRegexState := common.UsernameRegexCheck(usernameExclusionPat)
+	usernameMatches := usernameRegexState.Matches(data)
+
+	passwordRegexState := common.PasswordRegexCheck(" ") // No explicit character exclusions by Snowflake for passwords
+	passwordMatches := passwordRegexState.Matches(data)
+
+	for _, accountMatch := range accountMatches {
+		if len(accountMatch) != 2 {
+			continue
+		}
+		resAccountMatch := strings.TrimSpace(accountMatch[1])
+
+		for _, resUsernameMatch := range usernameMatches {
+
+			for _, resPasswordMatch := range passwordMatches {
+				_, metPasswordRequirements := meetsSnowflakePasswordRequirements(resPasswordMatch)
+
+				if !metPasswordRequirements {
+					continue
+				}
+
+				uri := fmt.Sprintf("%s:%s@%s/%s", resUsernameMatch, resPasswordMatch, resAccountMatch, database)
+
+				s1 := detectors.Result{
+					DetectorType: detectorspb.DetectorType_Snowflake,
+					Raw:          []byte(resPasswordMatch),
+					ExtraData: map[string]string{
+						"account":  resAccountMatch,
+						"username": resUsernameMatch,
+					},
+				}
+
+				if verify {
+					_ = &gosnowflake.Config{
+						Account:  resAccountMatch,
+						User:     resUsernameMatch,
+						Password: resPasswordMatch,
+						Database: database,
+					}
+
+					// Open a connection to Snowflake
+					db, err := sql.Open("snowflake", uri) // Needs the snowflake driver from gosnowflake
+
+					if err != nil {
+						s1.VerificationError = fmt.Errorf("unable to open a connection to Snowflake %+v", err)
+					}
+					defer db.Close()
+
+					err = db.Ping()
+					if err != nil {
+						if strings.Contains(err.Error(), "Incorrect username or password was specified") {
+							s1.Verified = false
+						}
+					} else {
+						rows, err := db.Query(retrieveAllDatabasesQuery)
+						if err != nil {
+							s1.VerificationError = fmt.Errorf("unable to query Snowflake to enrich secret ExtraData %+v", err)
+						}
+						defer rows.Close()
+
+						var databases []string
+						for rows.Next() {
+							var name, createdOn, isDefault, isCurrent, origin, owner, comment, option, retentionTime, kind string
+							err := rows.Scan(&createdOn, &name, &isDefault, &isCurrent, &origin, &owner, &comment, &option, &retentionTime, &kind)
+							if err != nil {
+								s1.VerificationError = fmt.Errorf("unable to finish querying Snowflake to enrich secret ExtraData %+v", err)
+							}
+							databases = append(databases, name)
+						}
+						s1.ExtraData["databases"] = strings.Join(databases, ", ")
+						s1.Verified = true
```

Comment:
The presence of a verification error (which is possible above) signals that the detector wasn't able to determine whether the candidate credentials are valid. That's not really consistent with setting `Verified` to true (which means that we know that the credentials are valid) - the framework assumes that you won't combine those two things, so you shouldn't. I'm not sure what the best solution here is - maybe putting those error messages somewhere else in the `Result` payload?

(I know that making this fourth "verification state" mathematically possible but forbidding its use is kind of weird, but it's the price we pay for being able to migrate detectors gradually.)

Reply:
ahh - good point